### PR TITLE
count of etcd object should be limited to the specified resource

### PR DESCRIPTION
**Pull request by tkashem**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:
Currently `Count` includes keys from different resource(s): any key (belonging to a different resource) that has the specified key as a prefix will be included in the count.

Consider the following keys:
```
A: <storage-prefix>//foo.bar.io/machines
B: <storage-prefix>//foo.bar.io/machinesets
```
If we ask for the count of key A, the result will also include the keys from key B since key B shares the same prefix as key A.

Append a separator to mark the end of the key, this will exclude all other keys from a different resource that is a prefix of the specified key.

**Special notes for your reviewer**:
* I am wondering if this change has any ripple effects. Is there any code site that relies on the current behavior (prefix matching) of the `Count` function.

**Does this PR introduce a user-facing change?**:
```release-note
Fix etcd_object_counts metric reported by kube-apiserver
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go [hunk 1]
```diff
@@ -452,6 +452,19 @@ func getNewItemFunc(listObj runtime.Object, v reflect.Value) func() runtime.Obje
...
+
+	// Consider the following keys:
+	// A: <storage-prefix>//foo.bar.io/machines
+	// B: <storage-prefix>//foo.bar.io/machinesets
+	// If we ask for the count of key A, the result will also include the count for key B
+	// since key B shares the same prefix as key A.
+	//
+	// Add a separator to mark the end of the key here, this will exclude all other
+	// keys from a different resource that is a prefix of the specified key.
+	if len(key) > 0 && key[len(key) - 1] != '/' {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:464]: Great finding - those this deserves a test to ensure we won't regress in the future.

Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:464]: Thanks @wojtek-t. I am working on a test. I wanted to make sure this is a good fix.

Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: @wojtek-t, just added a test, can you please take a look?

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go [hunk 2]
```diff
@@ -1818,6 +1818,13 @@ func testSetup(t *testing.T) (context.Context, *store, *integration.ClusterV3) {
 func testPropogateStore(ctx context.Context, t *testing.T, store *store, obj *example.Pod) (string, *example.Pod) {
 	// Setup store with a key and grab the output for returning.
 	key := "/testkey"
+	setOutput := testPropogateStoreWithKey(ctx, t, store, key, obj)
+	return key, setOutput
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go:1822]: nit:
```
return key, testPropagateStoreWithKey(ctx, t, store, key, obj)
```

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go [hunk 3]
```diff
@@ -2075,3 +2082,41 @@ func TestConsistentList(t *testing.T) {
...
+
+	resourceA := "/foo.bar.io/abc"
+
+	// resourceA is intentionally a prefix of resourceB to ensure that the count
+	// for resourceA does not include any objects from resourceB.
+	resourceB := fmt.Sprintf("%sdef", resourceA)
+
+	resourceACountExpected := 5
+	for i := 1; i <= resourceACountExpected; i++ {
+		obj := &example.Pod{ObjectMeta: metav1.ObjectMeta{Name: "foo"}}
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go:2098]: nit: can we ensure thse have unique names?

Name: fmt.Sprintf("foo-%d", i)

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go [hunk 4]
```diff
@@ -2075,3 +2082,41 @@ func TestConsistentList(t *testing.T) {
...
+
+		key := fmt.Sprintf("%s/%d", resourceA, i)
+		testPropogateStoreWithKey(ctx, t, store, key, obj)
+	}
+
+	resourceBCount := 4
+	for i := 1; i <= resourceBCount; i++ {
+		obj := &example.Pod{ObjectMeta: metav1.ObjectMeta{Name: "foo"}}
+
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go:2107]: same here

**Review by wojtek-t — COMMENTED** [review]
Just some nits - thanks for fixing it!

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go [hunk 5]
```diff
@@ -452,6 +452,19 @@ func getNewItemFunc(listObj runtime.Object, v reflect.Value) func() runtime.Obje
...
+	// A: <storage-prefix>//foo.bar.io/machines
+	// B: <storage-prefix>//foo.bar.io/machinesets
+	// If we ask for the count of key A, the result will also include the count for key B
+	// since key B shares the same prefix as key A.
+	//
+	// Add a separator to mark the end of the key here, this will exclude all other
+	// keys from a different resource that is a prefix of the specified key.
+	if len(key) > 0 && key[len(key)-1] != '/' {
+		key = fmt.Sprintf("%s/", key)
+	}
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:466]: use the same code as List():

https://github.com/kubernetes/kubernetes/blob/f64f687fcfb22e5099fbcb990f3e6faa44446fff/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go#L548-L553

**Comment by wojtek-t** [comment]
/lgtm
/approve

I added release-note - this should be cherrypicked back once it is merged at head.

**Comment by tkashem** [comment]
@liggitt @wojtek-t feedback addressed, please take a look.

**Comment by tkashem** [comment]
> I added release-note - this should be cherrypicked back once it is merged at head.

Hi @wojtek-t, I will cherry pick for 1.19, 1.8 and 1.17. That should be enough, right?

**Pull request merged**